data/cache/impl.py

```diff
@@ -213,3 +219,80 @@ def retrieve(self, cache_key, loader, should_cache=is_not_none):
             logger.debug("Not caching loaded result for key %s: %s", cache_key.key, result)
 
         return result
+
+
+class RedisDataModelCache(DataModelCache):
+    """
+    Implementation of the data model cache backed by a Redis service.
+    """
+
+    def __init__(
+        self,
+        host="127.0.0.1",
+        port=6379,
+        password=None,
+        db=0,
+        ca_cert=None,
+        ssl=False,
+    ):
+        self.client = StrictRedis(
+            host=host,
+            port=port,
+            password=password,
+            db=db,
+            ssl_ca_certs=ca_cert,
+            ssl=ssl,
+        )
+
+    def retrieve(self, cache_key, loader, should_cache=is_not_none):
+        # TODO: We might want to have different behavior based on `cache_key` (using "sets" for `/tags/list`, single value for others...)
+        not_found = None
+        if self.client is not None:
+            logger.debug("Checking cache for key %s", cache_key.key)
+            try:
+                cached_result = self.client.get(cache_key.key)
+                cache_retrieves.inc()
+                if cached_result != not_found:
+                    cache_hits.inc()
+
+                    return json.loads(cached_result)
+            except RedisError as re:
+                logger.warning("Got exception when trying to retrieve key %s", cache_key.key)
+
+        logger.debug("Found no result in cache for key %s; calling loader", cache_key.key)
+        result = loader()
+        logger.debug("Got loaded result for key %s: %s", cache_key.key, result)
+        if self.client is not None and should_cache(result):
+            # NOTE: This assumes that the Redis defined in `DATA_MODEL_CACHE_CONFIG` is the same as `USER_EVENTS_REDIS`.
+            with GlobalLock(cache_key.key, lock_ttl=5):
+                try:
+                    logger.debug(
+                        "Caching loaded result for key %s with expiration %s: %s",
+                        cache_key.key,
+                        result,
+                        cache_key.expiration,
+                    )
+                    expires = (
+                        convert_to_timedelta(cache_key.expiration) if cache_key.expiration else None
+                    )
+                    self.client.set(
+                        cache_key.key,
+                        json.dumps(result),
+                        ex=int(expires.total_seconds()) if expires else None,
+                    )
+                    logger.debug(
+                        "Cached loaded result for key %s with expiration %s: %s",
+                        cache_key.key,
+                        result,
+                        cache_key.expiration,
+                    )
+                except LockNotAcquiredException:
```

Comment:
I'm not sure if this is going to catch it. I think the GlobalLock context manager needs to be inside the try block.